File: src/main/java/io/github/chaosdave34/kitpvp/companions/Companion.java
package io.github.chaosdave34.kitpvp.companions;

import io.github.chaosdave34.ghutils.utils.PDCUtils;
import io.github.chaosdave34.kitpvp.ExtendedPlayer;
import io.github.chaosdave34.kitpvp.KitPvp;
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomFollowOwnerGoal;
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomMeleeAttackGoal;
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomOwnerHurtByTarget;
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomOwnerHurtTarget;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.PathfinderMob;
import net.minecraft.world.entity.ai.attributes.AttributeInstance;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.entity.ai.goal.FloatGoal;
import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
import net.minecraft.world.level.Level;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_20_R3.CraftWorld;
import org.bukkit.craftbukkit.v1_20_R3.entity.CraftMob;
import org.bukkit.craftbukkit.v1_20_R3.util.CraftVector;
import org.bukkit.entity.Player;
import org.bukkit.metadata.FixedMetadataValue;

public abstract class Companion {
    protected final String name;
    protected final double attackDamage;

    public Companion(String name, int attackDamage) {
        this.name = name;
        this.attackDamage = attackDamage;
    }

    public abstract PathfinderMob createVanillaMob(Level world);

    public org.bukkit.entity.Mob createCompanion(Player owner) {
        PathfinderMob companion = createVanillaMob(((CraftWorld) owner.getLocation().getWorld()).getHandle());
        companion.setPos(CraftVector.toNMS(owner.getLocation().toVector()));

        companion.getAttributes().registerAttribute(Attributes.ATTACK_DAMAGE);
        AttributeInstance attribute = companion.getAttribute(Attributes.ATTACK_DAMAGE);
        if (attribute != null) attribute.setBaseValue(attackDamage);

        companion.setCustomName(Component.literal(name));
        companion.setCustomNameVisible(true);

        companion.goalSelector.removeAllGoals(goal -> true);

        companion.goalSelector.addGoal(1, new FloatGoal(companion));
        companion.goalSelector.addGoal(2, new CustomMeleeAttackGoal(companion, owner, 1, false));
        companion.goalSelector.addGoal(3, new CustomFollowOwnerGoal(companion, owner, 1.0D, 10.0F, 2.0F, false));

        companion.targetSelector.removeAllGoals(goal -> true);

        companion.targetSelector.addGoal(1, new CustomOwnerHurtByTarget(companion, owner));
        companion.targetSelector.addGoal(2, new CustomOwnerHurtTarget(companion, owner));
        companion.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(companion, net.minecraft.world.entity.player.Player.class, 10, true, true, livingEntity -> {
            if (ExtendedPlayer.from(owner).inSpawn()) return false;

            if (!livingEntity.getUUID().equals(owner.getUniqueId())) {
                if (livingEntity.getBukkitLivingEntity() instanceof Player player) {
                    return ExtendedPlayer.from(player).inGame();
                } else return true;
            }
            return false;
        }));

        CraftMob bukkitMob = companion.getBukkitMob();
        bukkitMob.setMetadata("companion", new FixedMetadataValue(KitPvp.INSTANCE, true));
        PDCUtils.setOwner(bukkitMob, owner.getUniqueId());

        return bukkitMob;
    }

    public PathfinderMob createDummyCompanion(Location location) {
        PathfinderMob companion = createVanillaMob(((CraftWorld) location.getWorld()).getHandle());
        companion.setPos(CraftVector.toNMS(location.toVector()));
        companion.setRot(location.getYaw(), location.getPitch());
        companion.setYHeadRot(location.getYaw());
        companion.setInvulnerable(true);
        companion.setNoAi(true);
        companion.setSilent(true);

        companion.setCustomName(Component.literal(name));
        companion.setCustomNameVisible(true);

        return companion;
    }

}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/companions/Companion.kt
package io.github.chaosdave34.kitpvp.companions

import io.github.chaosdave34.kitpvp.ExtendedPlayer
import io.github.chaosdave34.kitpvp.KitPvp
import io.github.chaosdave34.kitpvp.extensions.setPDCOwner
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomFollowOwnerGoal
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomMeleeAttackGoal
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomOwnerHurtByTarget
import io.github.chaosdave34.kitpvp.pathfindergoals.CustomOwnerHurtTarget
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.PathfinderMob
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.ai.goal.FloatGoal
import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal
import net.minecraft.world.level.Level
import org.bukkit.Location
import org.bukkit.craftbukkit.CraftWorld
import org.bukkit.craftbukkit.util.CraftVector
import org.bukkit.entity.Player
import org.bukkit.metadata.FixedMetadataValue
import java.util.*
import net.minecraft.world.entity.player.Player as NMSPlayer

abstract class Companion(protected val name: String, private val attackDamage: Double) {

    abstract fun createVanillaMob(world: Level): PathfinderMob

    fun createCompanion(owner: Player): org.bukkit.entity.LivingEntity {
        val companion = createVanillaMob((owner.location.world as CraftWorld).handle)
        companion.setPos(CraftVector.toNMS(owner.location.toVector()))

        companion.attributes.registerAttribute(Attributes.ATTACK_DAMAGE)
        companion.getAttribute(Attributes.ATTACK_DAMAGE)?.baseValue = attackDamage

        companion.customName = Component.literal(name)
        companion.isCustomNameVisible = true

        companion.goalSelector.removeAllGoals { true }

        companion.goalSelector.addGoal(1, FloatGoal(companion))
        companion.goalSelector.addGoal(2, CustomMeleeAttackGoal(companion, owner, 1.0, false))
        companion.goalSelector.addGoal(3, CustomFollowOwnerGoal(companion, owner, 1.0, 10.0f, 2.0f, false))

        companion.targetSelector.removeAllGoals { true }

        companion.targetSelector.addGoal(1, CustomOwnerHurtByTarget(companion, owner))
        companion.targetSelector.addGoal(2, CustomOwnerHurtTarget(companion, owner))
        companion.targetSelector.addGoal(
            3,
            NearestAttackableTargetGoal(companion, NMSPlayer::class.java, 10, true, true)
            { livingEntity: LivingEntity -> checkNearestAttackableTargetGoal(livingEntity, owner.uniqueId) }
        )

        val bukkitMob = companion.bukkitLivingEntity
        bukkitMob.setMetadata("companion", FixedMetadataValue(KitPvp.INSTANCE, true))
        bukkitMob.setPDCOwner(owner)

        return bukkitMob
    }

    private fun checkNearestAttackableTargetGoal(livingEntity: LivingEntity, ownerUUID: UUID): Boolean {
        if (livingEntity.uuid == ownerUUID) return false

        val player = livingEntity.bukkitLivingEntity
        return if (player is Player) {
            ExtendedPlayer.from(player).inGame()
        } else true
    }

    fun createDummyCompanion(location: Location): PathfinderMob {
        val companion = createVanillaMob((location.world as CraftWorld).handle)
        companion.setPos(CraftVector.toNMS(location.toVector()))
        companion.setRot(location.yaw, location.pitch)
        companion.setYHeadRot(location.yaw)
        companion.isInvulnerable = true
        companion.isNoAi = true
        companion.isSilent = true

        companion.customName = Component.literal(name)
        companion.isCustomNameVisible = true

        return companion
    }
}